### ISSUE TITLE
Regression in type inference when using list comprehension in a loop

### ISSUE THREAD

**Issue by Molkree** [microsoft/pyright#4459]
**Describe the bug**
Pyright can't infer correct type if I use list comprehension in a loop. Regression introduced in [1.1.286](https://github.com/microsoft/pyright/releases/tag/1.1.286).

**Expected behavior**
No errors reported.

**Screenshots or Code**
```python
lst = [1]  # no error if I add explicit type annotation here (lst: list[int] = [1])
while True:
    lst = [val for val in lst]
```

![image](https://user-images.githubusercontent.com/12860284/212503010-2e3feb52-8301-4755-9316-e2c8540167dd.png)

**VS Code extension or command-line**
VSC Pylance since v2023.1.11 and up to v2023.1.21, VSC Pyright since 1.1.286 and up to 1.1.289.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
